From ad138a4a028bfa409acce82cebb91b916b79ddda Mon Sep 17 00:00:00 2001
From: Dan Liang <dan.liang@intel.com>
Date: Tue, 30 Apr 2013 14:13:27 +0800
Subject: Extend MPEGExtractor to support "qt" type

BZ: 48862

Original patches:
R3: 30130
R2: 19843

1. Check if the major brand is 'qt ' in box 'ftyp' to support quicktime mp4 clips;
2. If there are some invaild data at the end of one box, skip them and keep parsing.

Category: aosp improvement
Domain: Video.Media-container
Origin: internal
Upstream-Candidate: yes

Orig-Change-Id: I9ca5f14d7fd559d0eaabcac1c4559dd6d830f0c3
Change-Id: I8c7d8e355a63c6a999184afc3c6ea9835e24bb7e
Signed-off-by: Weian Chen <weian.chen@intel.com>
Signed-off-by: Dan Liang <dan.liang@intel.com>
Signed-off-by: Yuanjun Huang <yuanjun.huang@intel.com>
---
 media/libstagefright/MPEG4Extractor.cpp |    7 +++++--
 1 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/media/libstagefright/MPEG4Extractor.cpp b/media/libstagefright/MPEG4Extractor.cpp
index 94fb356..02137ce 100644
--- a/media/libstagefright/MPEG4Extractor.cpp
+++ b/media/libstagefright/MPEG4Extractor.cpp
@@ -755,7 +755,9 @@ status_t MPEG4Extractor::parseChunk(off64_t *offset, int depth) {
         }
     } else if (chunk_size < 8) {
         // The smallest valid chunk is 8 bytes long.
-        return ERROR_MALFORMED;
+        // Ignore the invalid data and continue
+        *offset += 4;
+        return OK;
     }
 
     char chunk[5];
@@ -3601,7 +3603,8 @@ static bool LegacySniffMPEG4(
         || !memcmp(header, "ftyp3ge6", 8) || !memcmp(header, "ftyp3gg6", 8)
         || !memcmp(header, "ftypisom", 8) || !memcmp(header, "ftypM4V ", 8)
         || !memcmp(header, "ftypM4A ", 8) || !memcmp(header, "ftypf4v ", 8)
-        || !memcmp(header, "ftypkddi", 8) || !memcmp(header, "ftypM4VP", 8)) {
+        || !memcmp(header, "ftypkddi", 8) || !memcmp(header, "ftypM4VP", 8)
+        || !memcmp(header, "ftypqt  ", 8)) {
         *mimeType = MEDIA_MIMETYPE_CONTAINER_MPEG4;
         *confidence = 0.4;
 
-- 
1.7.4.1

